doc/french-conjugator.1.in
.\" $Id: french-conjugator.1.in,v 1.13 2016/11/20 07:26:34 sarrazip Exp $
.\" french-conjugator - Analyzer of conjugated French verbs
.\" verbiste - French conjugation system
.\" Copyright (C) 2003-2009 Pierre Sarrazin <http://sarrazip.com/>
.\"
.\" This program is free software; you can redistribute it and/or
.\" modify it under the terms of the GNU General Public License
.\" as published by the Free Software Foundation; either version 2
.\" of the License, or (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
.\" 02110-1301, USA.
.\"
.\"
.TH french-conjugator "1" "@MANUAL_DATE_EN@" "" ""
.SH NAME
french-conjugator - conjugate French verbs
.SH SYNOPSIS
.B echo aimer | french-conjugator > result.txt
.SH DESCRIPTION
\fBfrench-conjugator\fR
reads the infinitive form of French verbs from
the command line or from standard input
and writes (to standard output) the complete conjugation of those
verbs, if they are known.
.PP
The standard input is not read if verbs are passed as command-line
arguments.
.PP
Each mode and tense is introduced by a line that starts with a hyphen
and a space, and ends with a colon.
The mode and tense in that line are always in English, regardless
of the user's current locale.
(This is meant to facilitate automatic parsing of the output.
For a French user interface, see the GNOME application and applet.)
The conjugation is ended with a line that only contains a hyphen.
If the given verb is unknown or not in the infinitive form,
only this line is written.
.PP
The command flushes its output buffer after finishing each answer.
This allows the command to be easily called from another program
through two pipes.
.PP
The command starts by loading its database from XML files
(stored typically in /usr/share/verbiste-0.1).
This takes some time, so it is a good idea to have the command
answer many requests instead of running it for each request.
.PP
The verbiste library's source archive contains Perl and Java
example programs that illustrate this technique.
.PP
There must not be any leading or trailing white spaces on the lines read
by the command.
.PP
In the \fBpast participle\fR tense, four lines are written: they correspond
in order
to the masculine singular, masculine plural, feminine singular and
feminine plural.
.SH OPTIONS
.TP
\fB\-\-help\fR
display a help page and exit
.TP
\fB\-\-version\fR
display version information and exit
.TP
\fB\-\-lang=L\fR
select the language to use (\fBfr\fR for French or \fBit\fR for Italian);
French is the default language
.TP
\fB\-\-mode=M\fR
only display mode \fBM\fR, where \fBM\fR can be
\fBinfinitive\fR,
\fBindicative\fR,
\fBconditional\fR,
\fBsubjunctive\fR,
\fBimperative\fR or
\fBparticiple\fR
(\fBM\fR can also be a comma-separated list of modes)
.TP
\fB\-\-tense=T\fR
only display tense \fBT\fR, where \fBT\fR can be
\fBpresent\fR,
\fBpast\fR,
\fBimperfect\fR or
\fBfuture\fR
(\fBT\fR can also be a comma-separated list of tenses)
.TP
\fB\-\-pronouns\fR
show the pronouns
.TP
\fB\-\-all-infinitives\fR
print the infinitive form of all the verbs in the knowledge base,
one per line, unsorted; other command-line arguments are ignored
.TP
\fB\-\-data-dir=D\fR
Get the XML data files from directory D instead of the default one.
.SH EXAMPLES
$ french-conjugator aimer
.br
- infinitive present:
.br
aimer
.br
- indicative present:
.br
aime
.br
aimes
.br
aime
.br
aimons
.br
aimez
.br
aiment
.br
- indicative imperfect:
.br
aimais
.br
aimais
.br
aimait
.br
[...]
.br
- participle past:
.br
aim\('e
.br
aim\('es
.br
aim\('ee
.br
aim\('ees
.br
-
.br
.SH LICENSE
This program is free software; you may redistribute it under the terms of
the GNU General Public License.  This program has absolutely no warranty.
.SH AUTHOR
See the verbiste(3) manual page.
.SH BUGS
See the verbiste(3) manual page.
.SH SEE ALSO
verbiste(3), french-deconjugator(1).


NEWS
2019-09-30: Version 0.1.47 released

    Fixed typos in the models for verbs léser, crémer, recéper
    and receper.

    Fixed wrong conjugation model for verb référencier.

    Improved the portability of the "make check" commands, namely for
    versions of make that start commands with the shell's -e option.

2019-07-28: Version 0.1.46 released

    The graphical interface can now be displayed in Italian, thanks
    to a contribution from a user.

    Fixed the HTML and LaTeX output.

    Fixed a bug where the graphical interface crashed after Escape was
    pressed in the Save As dialog.

2017-10-09: Version 0.1.45 released

    Dynamic C++ exception specifications removed, to comply with C++11.

2016-11-20: Version 0.1.44 released

    Fixed the conjugation of verb croître. Added verb recroître.

    Fixed a bug where french-conjugator would crash when an
    invalid mode or tense would be passed to --mode or --tense.

    Command french-conjugator now has options --mode and --tense:
    they accept a comma-separated list of modes or tenses respectively.

    Commands french-conjugator and french-deconjugator now have
    option --data-dir which allows the user to specify an XML data
    directory other than the default one.

    Removed the reference to -lstdc++ from the `examples' directory's
    Makefile.am, to resolve linking failures under GCC 5.3 and 6.1.

2016-03-21: Version 0.1.43 released

    Added verb abrévier.

    Fixed the plural second person (vous) of indicative and
    imperative present tenses for verbs contredire, dédire and
    prédire: e.g., prédisez is allowed in addition to prédites.

    Added a few Italian verbs and fixed some existing ones.
    Fixed some strings in the German graphical user interface.

2015-09-13: Version 0.1.42 released

    Fixed two memory leaks.
    Added Italian verbs.
    Added keywords to the GNOME .desktop file.

2014-04-06: Version 0.1.41 released

    Added support to build an applet for the MATE Desktop Environment
    (http://mate-desktop.org/).
    Added "cheyant" as an alternative present participle of the
    verb "choir":
        "Tout en cheyant je n’ai pu faire choix
        De mon point d’arrivée, -et j’ignore où je chois!"
        — (Edmond Rostand, "Cyrano de Bergerac", act III, scene 13)
    Fixed conjugation of verb sevrer.
    In the graphical application, the name of the verb that serves
    as the model for the displayed conjugation is now shown with
    that conjugation.
    The link in a conjugation page that normally points to the
    Wiktionnaire now points to the Wikizionario when the verb is Italian.
    Small source code fixes for portability to Solaris 10 and 11.
    The C example programs now fail properly if the Verbiste library
    fails to initialize.

2013-10-31: Version 0.1.40 released

    PNG icons of several sizes, which had been introduced in 0.1.39,
    are now pre-generated so that building Verbiste does not require
    the rsvg-convert command.
    Updated the config.guess and config.sub files to the 2013-06-10
    and 2013-10-01 versions respectively.
    Modernized the use of AC_INIT and AM_INIT_AUTOMAKE in configure.ac,
    which also now uses AC_CONFIG_SRCDIR.
    The configure script now builds the GNOME and GTK+ applications,
    by default, so that the distcheck target checks the src/gnome
    and src/gtk directories properly.

2013-10-28: Version 0.1.39 released

    Fixed conjugation of verbs blasphémer, brumasser, crémer,
    écrémer and luire.
    Removed bogus verb satisfer.
    Fixed manual pages which still mentioned obsolete command-line
    option --utf8.
    Replaced PNG icon with SVG one. PNG icons of several sizes
    are now generated by the install target.

2013-06-28: Version 0.1.38 released

    A conjugation page now shows a link to the Wiktionnaire page
    that is expected to contain the definition of the displayed verb.
    INSTALL and TODO files are not installed anymore.

2013-05-20: Version 0.1.37 released

    A verb can now be passed on the command-line of the GNOME and
    GTK+ applications to have their window display that verb's
    conjugation automatically.
    Updated the manual pages.

2013-04-21: Version 0.1.36 released

    Fixed conjugation of verb "quérir", which is only conjugated
    in the infinitive, unlike "acquérir".
    Fixed conjugation of verb "amu:ïr", which always takes an umlaut.
    Fixed conjugation of verbs "sécher", "assécher" and "déssécher"
    in the indicative future and conditional present tenses.
    Added verb "référencier".
    Fixed generate-all-forms.pl example for UTF-8 and so that it
    does not generate a temporary file anymore. It can still print
    only the infinitive forms if requested with the --inf option.

2012-11-18: Version 0.1.35 released

    Fixed a memory leak where the Trie<T> destructor omitted
    the destructon of the empty key user data.
    Manual pages: replaced UTF-8 characters with ASCII escape
    sequences.

2012-04-28: Version 0.1.34 released

    GNOME Applet is not packaged in the default RPM anymore.
    Added French verbs compoter, concaténer, occire and procrastiner.
    Created new conjugation model for verb harceler, which can be
    conjugated like "appeler" as well as like "peler".
    The user can now add verbs to Verbiste by creating an XML file
    called $HOME/.verbiste/verbs-<language>.xml. See the README or
    LISEZMOI file.

2011-04-03: Version 0.1.33 released

    Fixed errors with verb "chauvir".
    Added French verb "diligenter".
    The command-line tools do not accept Latin-1 input anymore
    (only UTF-8 is accepted now).
    Verbiste now works in UTF-8 and wide characters internally.
    The API has lost its Latin-1 conversion functions.
    RPM spec file rewritten using Fedora template.
    Added missing BuildRequires statement for perl-XML-Parser
    in RPM spec file.

2010-12-19: Version 0.1.32 released


src/commands/french-conjugator.cpp
/*  $Id: french-conjugator.cpp,v 1.34 2017/08/29 01:47:43 sarrazip Exp $
    french-conjugator.cpp - Conjugation of French verbs

    verbiste - French conjugation system
    Copyright (C) 2003-2010 Pierre Sarrazin <http://sarrazip.com/>

    This program is free software; you can redistribute it and/or
    modify it under the terms of the GNU General Public License
    as published by the Free Software Foundation; either version 2
    of the License, or (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
    02110-1301, USA.
*/


#include "Command.h"

#include "gui/conjugation.h"

#ifdef ENABLE_NLS
#ifdef HAVE_GETOPT_LONG
#include <unistd.h>
#ifndef _GNU_SOURCE
#define _GNU_SOURCE
#endif
#include <getopt.h>
#endif
#endif  /* ENABLE_NLS */

#include <langinfo.h>

#include <iostream>
#include <vector>
#include <algorithm>
#include <string.h>

using namespace std;
using namespace verbiste;


static const char *commandName = "french-conjugator";


#if defined(ENABLE_NLS) && defined(HAVE_GETOPT_LONG)
static struct option knownOptions[] =
{
    { "help",           no_argument,            NULL, 'h' },
    { "version",        no_argument,            NULL, 'v' },
    { "lang",           required_argument,      NULL, 'l' },
    { "mode",           required_argument,      NULL, 'm' },
    { "tense",          required_argument,      NULL, 't' },
    { "template",       required_argument,      NULL, 'e' },
    { "pronouns",       no_argument,            NULL, 'p' },
    { "all-infinitives",no_argument,            NULL, 'i' },
    { "data-dir",       required_argument,      NULL, 'd' },

    { NULL, 0, NULL, 0 }  // marks the end
};
#endif


class ConjugatorCommand : public Command
{
public:

    ConjugatorCommand(const string &conjugationFilename,
                      const string &verbsFilename,
                      FrenchVerbDictionary::Language lang)
      : Command(conjugationFilename, verbsFilename, lang),
        reqModes(),
        reqTenses(),
        reqTemplate(),
        includePronouns(false),
        aspirateH(false)
    {
        reqModes.reserve(8);
        reqTenses.reserve(8);
    }

    virtual ~ConjugatorCommand()
    {
    }

    // Constraints:
    vector<Mode>  reqModes;   // empty means all
    vector<Tense> reqTenses;  // empty means all
    string reqTemplate;
    bool includePronouns;

protected:

    virtual void processInputWord(const std::string &inputWord);

private:

    bool aspirateH;

    void displayTense(FrenchVerbDictionary &fvd,
                        const string &radical,
                        const TemplateSpec &templ,
                        Mode mode,
                        Tense tense);

    void displayConjugation(FrenchVerbDictionary &fvd,
                        const string &infinitive,
                        const string &tname,
                        const TemplateSpec &templ);
};


template <typename ElementType>
static bool
applyConstraint(const vector<ElementType> &v, ElementType candidateValue)
{
    if (v.empty())
        return true;  // no constraint, so always accept
    return find(v.begin(), v.end(), candidateValue) != v.end();
}


void
ConjugatorCommand::displayTense(FrenchVerbDictionary &fvd,
                                const string &radical,
                                const TemplateSpec &templ,
                                Mode mode,
                                Tense tense)
{
    if (!applyConstraint(reqModes, mode))
        return;
    if (!applyConstraint(reqTenses, tense))
        return;

    cout << "- " << FrenchVerbDictionary::getModeName(mode)
                << " " << FrenchVerbDictionary::getTenseName(tense) << ":\n";


    typedef vector<string> VS;
    typedef vector<VS> VVS;

    VVS conjug;
    fvd.generateTense(radical, templ, mode, tense, conjug,
                      includePronouns, aspirateH, lang == FrenchVerbDictionary::ITALIAN);

    for (VVS::const_iterator p = conjug.begin(); p != conjug.end(); p++)
    {
        for (VS::const_iterator i = p->begin(); i != p->end(); i++)
        {
            if (i != p->begin())
                cout << ", ";
            string s = *i;

            cout << s;
        }
        cout << "\n";
    }
}


void
ConjugatorCommand::displayConjugation(FrenchVerbDictionary &fvd,
                                        const string &infinitive,
                                        const string &tname,
                                        const TemplateSpec &templ)
{
    try
    {
        aspirateH = fvd.isVerbStartingWithAspirateH(infinitive);

        string radical = FrenchVerbDictionary::getRadical(infinitive, tname);

        if (lang == FrenchVerbDictionary::FRENCH || lang == FrenchVerbDictionary::ITALIAN)
        {
            displayTense(fvd, radical, templ, INFINITIVE_MODE, PRESENT_TENSE);

            displayTense(fvd, radical, templ, INDICATIVE_MODE, PRESENT_TENSE);
            displayTense(fvd, radical, templ, INDICATIVE_MODE, IMPERFECT_TENSE);
            displayTense(fvd, radical, templ, INDICATIVE_MODE, FUTURE_TENSE);
            displayTense(fvd, radical, templ, INDICATIVE_MODE, PAST_TENSE);

            displayTense(fvd, radical, templ, CONDITIONAL_MODE, PRESENT_TENSE);

            displayTense(fvd, radical, templ, SUBJUNCTIVE_MODE, PRESENT_TENSE);
            displayTense(fvd, radical, templ, SUBJUNCTIVE_MODE, IMPERFECT_TENSE);

            displayTense(fvd, radical, templ, IMPERATIVE_MODE, PRESENT_TENSE);

            displayTense(fvd, radical, templ, PARTICIPLE_MODE, PRESENT_TENSE);
            displayTense(fvd, radical, templ, PARTICIPLE_MODE, PAST_TENSE);
        }
        if (lang == FrenchVerbDictionary::ITALIAN)
            displayTense(fvd, radical, templ, GERUND_MODE, PRESENT_TENSE);

        if (lang == FrenchVerbDictionary::GREEK)
        {
            displayTense(fvd, radical, templ, PRESENT_INDICATIVE, ACTIVE_TENSE);
            displayTense(fvd, radical, templ, PRESENT_INDICATIVE, PASSIVE_TENSE);
            displayTense(fvd, radical, templ, PRESENT_SUBJUNCTIVE, ACTIVE_TENSE);
            displayTense(fvd, radical, templ, PRESENT_SUBJUNCTIVE, PASSIVE_TENSE);
            displayTense(fvd, radical, templ, PRESENT_IMPERATIVE, IMPERATIVE_ACTIVE_TENSE);
            displayTense(fvd, radical, templ, PRESENT_IMPERATIVE, IMPERATIVE_PASSIVE_TENSE);
            displayTense(fvd, radical, templ, PRESENT_GERUND, PRESENT_TENSE);
            displayTense(fvd, radical, templ, PAST_IMPERFECT_INDICATIVE, ACTIVE_TENSE);
            displayTense(fvd, radical, templ, PAST_IMPERFECT_INDICATIVE, PASSIVE_TENSE);
            displayTense(fvd, radical, templ, PAST_PERFECT_INDICATIVE, ACTIVE_TENSE);
            displayTense(fvd, radical, templ, PAST_PERFECT_INDICATIVE, PASSIVE_TENSE);
            displayTense(fvd, radical, templ, PAST_PERFECT_SUBJUNCTIVE, ACTIVE_TENSE);
            displayTense(fvd, radical, templ, PAST_PERFECT_SUBJUNCTIVE, PASSIVE_TENSE);
            displayTense(fvd, radical, templ, PAST_PERFECT_IMPERATIVE, IMPERATIVE_ACTIVE_TENSE);
            displayTense(fvd, radical, templ, PAST_PERFECT_IMPERATIVE, IMPERATIVE_PASSIVE_TENSE);
            displayTense(fvd, radical, templ, PAST_PERFECT_INFINITIVE, PAST_PERFECT);
        }
    }
    catch (logic_error &e)
    {
        return;
    }
}


/*virtual*/
void
ConjugatorCommand::processInputWord(const string &inputWord)
{
    const char *tname = (reqTemplate.empty() ? NULL : reqTemplate.c_str());
    const std::set<std::string> *templateSet = NULL;
    std::set<std::string> singleton;

    if (tname == NULL)  // if no specific template requested
    {
        // Use the template associated with the verb, if known:
        templateSet = &fvd->getVerbTemplateSet(inputWord);
    }
    else
    {
        singleton.insert(tname);
        templateSet = &singleton;
    }

    for (std::set<std::string>::const_iterator it = templateSet->begin();
                                               it != templateSet->end(); ++it)
    {
        if (it != templateSet->begin())
            cout << "-\n";  // separate conjugations

        const string &tname = *it;
        const TemplateSpec *templ = fvd->getTemplate(tname);

        if (templ != NULL)
            displayConjugation(*fvd, inputWord, tname, *templ);
    }

    cout << "-" << endl;  // marks the end of the answer, and flushes it
}


#if defined(ENABLE_NLS) && defined(HAVE_GETOPT_LONG)

static
void
displayVersionNo()
{
    cout << commandName << ' ' << VERSION << '\n';
}


static
void
displayHelp()
{
    cout << '\n';

    displayVersionNo();

    cout << "Part of " << PACKAGE << " " << VERSION << "\n";

    cout <<
"\n"
"Copyright (C) " COPYRIGHT_YEARS " Pierre Sarrazin <http://sarrazip.com/>\n"
"This program is free software; you may redistribute it under the terms of\n"
"the GNU General Public License.  This program has absolutely no warranty.\n"
    ;

    cout <<
"\n"
"Options:\n"
"--help             Display this help page and exit\n"
"--version          Display this program's version number and exit\n"
"--lang=L           Select language L (fr for French, it for Italian)\n"
"                   Default is French.\n"
"--mode=M           Only display mode M (infinitive, indicative, etc.)\n"
"                   (M can also be a comma-separated list of modes)\n"
"--tense=T          Only display tense T (present, past, etc)\n"
"                   (T can also be a comma-separated list of tenses)\n"
"--template=T       Use template T to conjugate the verbs\n"
"--pronouns         Include pronouns in the displayed conjugation\n"
"--all-infinitives  Print the names of all known verbs, one per line (unsorted)\n"
"--data-dir=D       Get XML data files from directory D instead of default\n"
"                   directory " << LIBDATADIR << "\n"
"\n"
"See the " << commandName << "(1) manual page for details.\n"
"\n"
    ;
}

#endif


static vector<string>
split(char delimiter, const char *input)
{
    vector<string> parts;
    for (;;)
    {
        const char *foundDelim = strchr(input, delimiter);
        if (!foundDelim)
        {
            if (*input)
                parts.push_back(input);
            return parts;
        }
        if (foundDelim != input)  // if part not empty
        {
            parts.push_back(string(input, 0, foundDelim - input));
        }
        input = foundDelim + 1;
    }
}


int
main(int argc, char *argv[])
{
    setlocale(LC_CTYPE, "");

    vector<Mode> reqModes;
    reqModes.reserve(8);
    vector<Tense> reqTenses;
    reqTenses.reserve(8);
    string reqTemplate;
    bool includePronouns = false;
    bool listAllInfinitives = false;
    string langCode = "fr";

    #if defined(ENABLE_NLS) && defined(HAVE_GETOPT_LONG)

    /*  Interpret the command-line options:
    */
    int c;
    try
    {
        do
        {
            c = getopt_long(argc, argv, "hvm:t:l:p", knownOptions, NULL);

            switch (c)
            {
                case EOF:
                    break;  // nothing to do

                case 'l':
                    langCode = optarg;
                    if (FrenchVerbDictionary::parseLanguageCode(langCode) == FrenchVerbDictionary::NO_LANGUAGE)
                    {
                        cerr << commandName << ": invalid language code " << optarg << "\n";
                        return EXIT_FAILURE;
                    }
                    break;

                case 'm':
                {
                    // Split the argument on commas, check that each mode name is valid,
                    // accumulate modes in reqModes[].
                    //
                    vector<string> reqModeNames = split(',', optarg);
                    for (vector<string>::const_iterator it = reqModeNames.begin();
                                                       it != reqModeNames.end(); ++it)
                    {
                        const string &modeName = *it;
                        Mode reqMode = FrenchVerbDictionary::convertModeName(modeName.c_str());
                        if (reqMode == INVALID_MODE)
                            cerr << commandName << ": invalid mode " << modeName << "\n";
                        else
                            reqModes.push_back(reqMode);
                    }
                    if (reqModes.size() != reqModeNames.size())
                        return EXIT_FAILURE;
                    break;
                }

                case 't':
                {
                    // Similar to 'm'.
                    //
                    vector<string> reqTenseNames = split(',', optarg);
                    for (vector<string>::const_iterator it = reqTenseNames.begin();
                                                       it != reqTenseNames.end(); ++it)
                    {
                        const string &tenseName = *it;
                        Tense reqTense = FrenchVerbDictionary::convertTenseName(tenseName.c_str());
                        if (reqTense == INVALID_TENSE)
                            cerr << commandName << ": invalid tense " << tenseName << "\n";
                        else
                            reqTenses.push_back(reqTense);
                    }
                    if (reqTenses.size() != reqTenseNames.size())
                        return EXIT_FAILURE;
                    break;
                }

                case 'e':
                    reqTemplate = optarg;
                    break;

                case 'p':
                    includePronouns = true;
                    break;

                case 'v':
                    displayVersionNo();
                    return EXIT_SUCCESS;

                case 'h':
                    displayHelp();
                    return EXIT_SUCCESS;

                case 'i':
                    listAllInfinitives = true;
                    break;

                case 'd':
                    if (optarg[0] == '\0')
                    {
                        cerr << commandName << ": invalid data directory given to --data-dir\n";
                        return EXIT_FAILURE;
                    }
                    if (setenv("LIBDATADIR", optarg, true) != 0)  // for FrenchVerbDictionary::getXMLFilenames()
                    {
                        int e = errno;
                        cerr << commandName << ": failed to use the data directory ("
                             << optarg << ") given to --data-dir: " << strerror(e) << "\n";
                        return EXIT_FAILURE;
                    }
                    break;

                default:
                    displayHelp();
                    return EXIT_FAILURE;
            }
        } while (c != EOF && c != '?');
    }
    catch (const exception &e)
    {
        cerr << commandName << ": exception: " << e.what() << endl;
        return EXIT_FAILURE;
    }

    #else

    int optind = 1;

    #endif

    try
    {
        FrenchVerbDictionary::Language lang = FrenchVerbDictionary::parseLanguageCode(langCode);
        if (lang == FrenchVerbDictionary::NO_LANGUAGE)
            lang = FrenchVerbDictionary::FRENCH;

        string conjFN, verbsFN;
        FrenchVerbDictionary::getXMLFilenames(conjFN, verbsFN, lang);

        ConjugatorCommand cmd(conjFN, verbsFN, lang);

        if (listAllInfinitives)
            return cmd.listAllInfinitives(cout);

        cmd.reqModes = reqModes;
        cmd.reqTenses = reqTenses;
        cmd.reqTemplate = reqTemplate;
        cmd.includePronouns = includePronouns;

        if (!reqTemplate.empty() &&
                    cmd.getFrenchVerbDictionary().getTemplate(reqTemplate) == NULL)
        {
            cerr << commandName << ": invalid conjugation template "
                                                << optarg << "\n";
            return EXIT_FAILURE;
        }

        return cmd.run(argc - optind, argv + optind);
    }
    catch (const exception &e)
    {
        cerr << commandName << ": exception: " << e.what() << endl;
        return EXIT_FAILURE;
    }
}
